### PR TITLE
VACMS-14289: Enhance magichead error messaging

### PR DESCRIPTION
**Pull request by dsasser**
## Description
Error message display improvements to magichead.

This PR:

- Moves the magichead 'max depth' setting to the field instance setting rather than a field widget setting. This allowed for a cleaner implementation overall than keeping the setting as a field widget setting.
- Moved magichead constraints to the field item list class, rather than the field item. This allowed for removing some ugly static caching as well as supporting multiple magichead fields on the same entity/page/node.
- Adds new validations
- Adds a javascript library that ensures that magichead errors are always visible

Relates to #14289

## Testing done
Tested using VA Benefits Taxonomy's magichead field.

## Screenshots

Screenshot 2023-08-25 at 8 37 07 AM
Screenshot 2023-08-25 at 8 36 16 AM
Screenshot 2023-08-25 at 8 33 58 AM

## QA steps

What needs to be checked to prove this works?
What needs to be checked to prove it didn't break any related things?
What variations of circumstances (users, actions, values) need to be checked?

Screenshot 2023-08-25 at 8 49 28 AM

As a content admin
1. Visit /admin/structure/taxonomy/manage/va_benefits_taxonomy/add
2. Populate 5 sections all at depth 0 (do not nest children under parents)
3. Click Save to save the form
4. Click Edit to edit the new term you just created
5. Organize the magichead such that they are organized as in the screenshot above
6. Click on 'show row weights' to enter accessible mode
7. Change the last section depth from 3 to 4
8. Save the form
   - [x] Validate you receive an error message at the top of the page displays: 1 error has been found: [Depth for row 5]
   - [x] Validate that clicking on [Depth for row 5] link scrolls you to the error
   - [x] Validate that the error displays: Maximum depth allowed is 3 
9. Change the depth from 4 to -4 (negative 4)
10. Save the form
   - [x] Validate you receive an error message at the top of the page displays: 1 error has been found: [Depth for row 5]
   - [x] Validate that clicking on [Depth for row 5] link scrolls you to the error
   - [x] Validate that the error displays: Depth must be a positive integer
11. Click 'hide row weights'
12. Save the form
   - [x] Validate you receive an error message at the top of the page displays: 1 error has been found: [Depth for row 5]
   - [x] Validate that clicking on [Depth for row 5] link scrolls you to the error
   - [x] Validate that the error is visible
13. Change the depth of section 5 back to 3
14. Change the depth of section 1 from 1 to 2
15. Save the form
   - [x] Validate you receive an error message at the top of the page displays: 1 error has been found: [Depth for row 2]
   - [x] Validate that clicking on [Depth for row 2] link scrolls you to the error
   - [x] Validate the error displays: Depth for a Section cannot be more than 1 greater than the Section immediately above it

### Definition of Done

- [ ] Documentation has been updated, if applicable.
- [ ] Tests have been added if necessary.
- [ ] Automated tests have passed.
- [ ] Code Quality Tests have passed.
- [ ] Acceptance Criteria in related issue are met.
- [ ] Manual Code Review Approved.

### Select Team for PR review

- [ ] `CMS Team`
- [x] `Public websites`
- [ ] `Facilities`
- [ ] `User support`
- [ ] `Accelerated Publishing`

### CODE REVIEW

**Comment by chri5tia** [comment]
I am noticing that something broke with the theme, where the dotted line is showing up in all tabledrop items now. It's on my local too so it must have merged before this PR.

Screenshot 2023-08-25 at 1 55 26 PM

**Comment by dsasser** [comment]
> I am noticing that something broke with the theme, where the dotted line is showing up in all tabledrop items now. It's on my local too, apparently, so it must have merged before this PR.
> 
> Screenshot 2023-08-25 at 1 55 26 PM

The scope of the selector is not specific to magichead fields. So any page with a magichead field will cause all tables to show the dotted lines.

```
.tabledrag-cell-content > * {
  border-left: #ccc 1px dashed;
}
```

**Comment by chri5tia** [comment]
Nice work

File: docroot/modules/custom/va_gov_magichead/js/magichead.example.es6.js [hunk 1]
```diff
@@ -0,0 +1,78 @@
+/**
+ * @file
+ * DO NOT USE IN PRODUCTION.
+ * These are example Magichead behaviors. Because it is highly likely we will
+ * need to adapt tableDrag core behavior in the future, I wanted to preserve
+ * knowledge I gained by studying the tableDrag.js API.
```
Comment (chri5tia) [inline docroot/modules/custom/va_gov_magichead/js/magichead.example.es6.js:6]: Should this still be here if it is merging?

File: docroot/modules/custom/va_gov_magichead/js/magichead.example.js [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+(function (Drupal) {
+  function isValidSwap() {
+    return true;
+  }
+  function onDrop() {
+    var dragObject = this;
+    var rowElement = dragObject.rowObject.element;
+    rowElement.classList.add("error");
+    var newElement = document.createElement("div");
+    newElement.textContent = "Max Depth Exceeded! 😡";
```
Comment (chri5tia) [inline docroot/modules/custom/va_gov_magichead/js/magichead.example.js:16]: Haha is this angry face emoji going to cause any problems?

File: docroot/modules/custom/va_gov_magichead/src/Plugin/Field/FieldType/MagicheadItem.php [hunk 3]
```diff
@@ -13,19 +14,34 @@
  *   category = @Translation("Reference revisions"),
  *   default_widget = "magichead_paragraphs_classic",
  *   default_formatter = "entity_reference_label",
- *   list_class = "\Drupal\entity_reference_hierarchy_revisions\EntityReferenceHierarchyRevisionsFieldItemList",
+ *   list_class = "\Drupal\va_gov_magichead\MagicheadFieldItemList",
  * )
  */
 class MagicheadItem extends EntityReferenceHierarchyRevisionsItem {
```
Comment (chri5tia) [inline docroot/modules/custom/va_gov_magichead/src/Plugin/Field/FieldType/MagicheadItem.php:20]: I'm wondering how this is working, to change the list class even though magicheadItem extends Entity Ref... mentioned in the changed line?

**Review by chri5tia — APPROVED** [review]

File: docroot/modules/custom/va_gov_magichead/src/Plugin/Field/FieldType/MagicheadItem.php (hunk 3 above)
Comment (dsasser) [reply docroot/modules/custom/va_gov_magichead/src/Plugin/Field/FieldType/MagicheadItem.php:20]: MagicheadItem, a field item class (think of this as a single field instance/delta), extends EntityReferenceHierarchyRevisionsItem. This hasn't changed. What I did here is change the class which is responsible fore building a _list_ of MagicheadItem items. This is defined in the field type annotation, and doesn't affect the class inheritance.

File: docroot/modules/custom/va_gov_magichead/js/magichead.example.js (hunk 2 above)
Comment (dsasser) [reply docroot/modules/custom/va_gov_magichead/js/magichead.example.js:16]: Ha I forgot I put this in. Yes the emoji is harmless, as any character set which supports utf-8 can display it, and if not, a □ or � would be displayed instead.

File: docroot/modules/custom/va_gov_magichead/js/magichead.example.es6.js (hunk 1 above)
Comment (dsasser) [reply docroot/modules/custom/va_gov_magichead/js/magichead.example.es6.js:6]: Yes I would like to leave it in. I went through a lot of trouble to figure out the entire rendering pipeline for tableDrag, and outlined how to interact with the API and would like to preserve that knowledge somewhere findable.

Are you proposing we leave it out?

Comment (chri5tia) [reply docroot/modules/custom/va_gov_magichead/js/magichead.example.es6.js:6]: I think it is great to have all these notes in there. I was only confused about the warning to not use it in production. I see that that header means don't use these example behaviors on prod. I like it.

**Pull request merged**